pkg/rpcserver/runner.go

```diff
@@ -321,21 +407,43 @@ func (runner *Runner) sendSignalUpdate(plus, minus []uint64) error {
 			},
 		},
 	}
-	return flatrpc.Send(runner.conn, msg)
+	return flatrpc.Send(conn, msg)
 }
 
-func (runner *Runner) sendCorpusTriaged() error {
+func (runner *Runner) SendCorpusTriaged() error {
+	runner.mu.Lock()
+	conn := runner.conn
+	runner.mu.Unlock()
+
+	if conn == nil {
+		panic("no handshake is done yet")
+	}
+
 	msg := &flatrpc.HostMessage{
 		Msg: &flatrpc.HostMessages{
 			Type:  flatrpc.HostMessagesRawCorpusTriaged,
 			Value: &flatrpc.CorpusTriaged{},
 		},
 	}
-	return flatrpc.Send(runner.conn, msg)
+	return flatrpc.Send(conn, msg)
 }
 
-func (runner *Runner) shutdown(crashed bool) ([]ExecRecord, []byte) {
-	if runner.conn != nil {
+func (runner *Runner) Stop() {
+	runner.mu.Lock()
+	runner.stopped = true
+	conn := runner.conn
+	runner.mu.Unlock()
+	if conn != nil {
+		conn.Close()
+	}
+}
+
+func (runner *Runner) Shutdown(crashed bool) []ExecRecord {
+	runner.mu.Lock()
+	conn := runner.conn
+	runner.mu.Unlock()
+
+	if conn != nil {
 		// Wait for the connection goroutine to finish and stop touching data.
 		// If conn is nil before we removed the runner, then it won't touch anything.
 		<-runner.finished
```

Comment:
This now can hang if Handshake has set runner.conn, but then we exited before calling runner.ConnectionLoop which will close runner.finished.

Reply:
Refactored the code to set `finished` in the connection loop.